File: Leetcode/0089-gray-code/solution.py
from typing import List


class Solution:
    def grayCode(self, n: int) -> List[int]:
        """Time O(2^n), Space O(2^n)."""
        return [i ^ (i >> 1) for i in range(1 << n)]


def validate(seq, n):
    if len(seq) != (1 << n): return False
    if seq[0] != 0: return False
    seen = set()
    for v in seq:
        if v < 0 or v >= (1 << n) or v in seen: return False
        seen.add(v)
    for i in range(len(seq)):
        j = (i + 1) % len(seq)
        if bin(seq[i] ^ seq[j]).count('1') != 1: return False
    return True


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    for n in [1, 2, 3, 4, 5, 8]:
        seq = sol.grayCode(n)
        if validate(seq, n):
            print(f"PASS: n={n}, len={len(seq)}"); passed += 1
        else:
            print(f"FAIL: n={n}, seq={seq}"); failed += 1
    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")

File: Leetcode/0089-gray-code/solution.go
package main

import (
	"fmt"
	"math/bits"
)

// 0089. Gray Code
// Time: O(2^n), Space: O(2^n)
func grayCode(n int) []int {
	size := 1 << n
	result := make([]int, size)
	for i := 0; i < size; i++ {
		result[i] = i ^ (i >> 1)
	}
	return result
}

func validate(seq []int, n int) bool {
	if len(seq) != 1<<n {
		return false
	}
	if seq[0] != 0 {
		return false
	}
	seen := make(map[int]bool)
	for _, v := range seq {
		if v < 0 || v >= 1<<n || seen[v] {
			return false
		}
		seen[v] = true
	}
	for i := 0; i < len(seq); i++ {
		j := (i + 1) % len(seq)
		if bits.OnesCount(uint(seq[i]^seq[j])) != 1 {
			return false
		}
	}
	return true
}

func main() {
	passed, failed := 0, 0
	for _, n := range []int{1, 2, 3, 4, 5, 8} {
		seq := grayCode(n)
		if validate(seq, n) {
			fmt.Printf("PASS: n=%d, len=%d\n", n, len(seq))
			passed++
		} else {
			fmt.Printf("FAIL: n=%d, seq=%v\n", n, seq)
			failed++
		}
	}
	fmt.Printf("\nResults: %d passed, %d failed, %d total\n", passed, failed, passed+failed)
}
